mnd/mnd/integration/mims/mims_cache.py

```diff
@@ -35,21 +35,40 @@ def _min_ts(input_dict):
 
 
 @func.ttl_cache(maxsize=1, ttl=3600)
-def _min_product_expiry_ts():
-    return _min_ts(MIMSProductCache.objects.aggregate(Min('expires_on')))
+def _min_search_expiry_ts():
+    return _min_ts(MIMSSearchTerm.objects.aggregate(Min('expires_on')))
 
 
 @func.ttl_cache(maxsize=1, ttl=3600)
 def _min_cmi_expiry_ts():
     return _min_ts(MIMSCmiCache.objects.aggregate(Min('expires_on')))
 
 
+def write_search_results(search_term, products):
+    existing = MIMSSearchTerm.objects.filter(search_term__iexact=search_term).first()
+    if existing:
+        if existing.expires_on <= timezone.now():
+            existing.products = products
+            existing.expires_on = _expiring_ts()
+            existing.save()
+    else:
+        MIMSSearchTerm.objects.create(
+            search_term=search_term,
+            products=products,
+            expires_on=_expiring_ts()
+        )
+
+
 def evict_expired_entries():
-    if (min_expiry := _min_product_expiry_ts()) and timezone.now() > min_expiry:
-        product_cache_qs = MIMSProductCache.objects.filter(expires_on__lt=timezone.now())
-        logger.info(f"Evicting {product_cache_qs.count()} product cache entries")
-        product_cache_qs.delete()
-        _min_product_expiry_ts.cache_clear()
+    if (min_expiry := _min_search_expiry_ts()) and timezone.now() > min_expiry:
+        search_term_qs = MIMSSearchTerm.objects.filter(expires_on__lt=timezone.now())
+        logger.info(f"About to evict {search_term_qs.count()} search term entries")
+        for st in search_term_qs:
+            MIMSProductCache.objects.filter(product_id__in=st.products).update(ref_count=F('ref_count') - 1)
+        search_term_qs.delete()
+        products_to_delete = MIMSProductCache.objects.filter(ref_count__lte=0)
+        logger.info(f"Deleting {products_to_delete.count()} products with no references")
+        products_to_delete.delete()
```

Comment:
Needs `_min_search_expiry_ts.cache_clear()` here as well to stop it from trying to delete products for the entire hour

Reply:
Updated.